File: 프로그래머스/4/1843. 사칙연산/사칙연산.py
INF = int(1e9)


def solution(arr):
    max_dp = [[-INF] * 202 for _ in range(202)]
    min_dp = [[INF] * 202 for _ in range(202)]
    return get_max(0, len(arr), arr, max_dp, min_dp)


def get_max(start, end, arr, max_dp, min_dp):
    if max_dp[start][end] != -INF:
        return max_dp[start][end]
    if end - start == 1:
        return int(arr[start])
    
    result = -INF
    
    for i in range(start + 1, end, 2):
        if arr[i] == '+':
            v1 = get_max(start, i, arr, max_dp, min_dp)
            v2 = get_max(i + 1, end, arr, max_dp, min_dp)
            v = v1 + v2
        else:
            v1 = get_max(start, i, arr, max_dp, min_dp)
            v2 = get_min(i + 1, end, arr, max_dp, min_dp)
            v = v1 - v2
        
        result = max(result, v)
    
    max_dp[start][end] = result
    return result


def get_min(start, end, arr, max_dp, min_dp):
    if min_dp[start][end] != INF:
        return min_dp[start][end]
    if end - start == 1:
        return int(arr[start])
    
    result = INF
    
    for i in range(start + 1, end, 2):
        if arr[i] == '+':
            v1 = get_min(start, i, arr, max_dp, min_dp)
            v2 = get_min(i + 1, end, arr, max_dp, min_dp)
            v = v1 + v2
        else:
            v1 = get_min(start, i, arr, max_dp, min_dp)
            v2 = get_max(i + 1, end, arr, max_dp, min_dp)
            v = v1 - v2
    
        result = min(result, v)
    
    min_dp[start][end] = result
    return result

File: 프로그래머스/4/1843. 사칙연산/사칙연산.java
import java.util.Arrays;

class Solution {
    private static final int[][] maxDp = new int[202][202];
    private static final int[][] minDp = new int[202][202];
    
    private static int max(int start, int end, String[] arr) {
        if (end - start == 1) {
            return Integer.parseInt(arr[start]);
        } 
        if (maxDp[start][end] > Integer.MIN_VALUE) {
            return maxDp[start][end];
        }
        
        for (int mid = start + 1; mid < end; mid += 2) {
            int v;
            
            if (arr[mid].equals("+")) {
                v = max(start, mid, arr) + max(mid + 1, end, arr);
            } else {
                v = max(start, mid, arr) - min(mid + 1, end, arr);
            }
            
            maxDp[start][end] = Math.max(maxDp[start][end], v);
        }
        
        return maxDp[start][end];
    }
    
    private static int min(int start, int end, String[] arr) {
        if (end - start == 1) {
            return Integer.parseInt(arr[start]);
        } 
        if (minDp[start][end] < Integer.MAX_VALUE) {
            return minDp[start][end];
        }
        
        for (int mid = start + 1; mid < end; mid += 2) {
            int v;

            if (arr[mid].equals("+")) {
                v = min(start, mid, arr) + min(mid + 1, end, arr);
            } else {
                v = min(start, mid, arr) - max(mid + 1, end, arr);
            }
            
            minDp[start][end] = Math.min(minDp[start][end], v);
        }

        return minDp[start][end];
    }
    
    public int solution(String arr[]) {
        for (int[] row : maxDp) Arrays.fill(row, Integer.MIN_VALUE);
        for (int[] row : minDp) Arrays.fill(row, Integer.MAX_VALUE);
        
        return max(0, arr.length, arr);
    }
}